src/core/pointcloud/qgsremoteeptpointcloudindex.cpp

```diff
@@ -0,0 +1,521 @@
+/***************************************************************************
+                         qgsremoteeptpointcloudindex.cpp
+                         --------------------
+    begin                : March 2021
+    copyright            : (C) 2021 by Belgacem Nedjima
+    email                : belgacem dot nedjima at gmail dot com
+ ***************************************************************************/
+
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgsremoteeptpointcloudindex.h"
+
+#include <QFile>
+#include <QFileInfo>
+#include <QDir>
+#include <QJsonArray>
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QTime>
+#include <QtDebug>
+#include <QQueue>
+#include <QTimer>
+
+#include "qgseptdecoder.h"
+#include "qgscoordinatereferencesystem.h"
+#include "qgspointcloudrequest.h"
+#include "qgspointcloudattribute.h"
+#include "qgslogger.h"
+#include "qgsfeedback.h"
+#include "qgsmessagelog.h"
+
+#include "qgstiledownloadmanager.h"
+#include "qgsblockingnetworkrequest.h"
+#include "qgsfiledownloader.h"
+
+#include "qgsfileutils.h"
+#include "qgsapplication.h"
+#include "qgspointcloudblockhandle.h"
+
+///@cond PRIVATE
+
+#define PROVIDER_KEY QStringLiteral( "ept" )
+#define PROVIDER_DESCRIPTION QStringLiteral( "EPT point cloud provider" )
+
+QgsRemoteEptPointCloudIndex::QgsRemoteEptPointCloudIndex() : QgsPointCloudIndex()
+{
+  mTileDownloadManager = QgsApplication::tileDownloadManager();
+}
+
+QgsRemoteEptPointCloudIndex::~QgsRemoteEptPointCloudIndex() = default;
+
+void QgsRemoteEptPointCloudIndex::load( const QString &url )
+{
+  mUrl = QUrl( url );
+
+  QStringList splitUrl = url.split( '/' );
+
+  mUrlFileNamePart = splitUrl.back();
+  splitUrl.pop_back();
+  mUrlDirectoryPart = splitUrl.join( '/' );
+
+  QNetworkRequest nr( url );
+
+  QgsBlockingNetworkRequest req;
+  QgsBlockingNetworkRequest::ErrorCode errCode = req.get( nr );
+  if ( errCode != QgsBlockingNetworkRequest::NoError )
+  {
+    QgsDebugMsg( QStringLiteral( "Request failed: " ) + url );
+    mIsValid = false;
+    return;
+  }
+
+  QgsNetworkReplyContent reply = req.reply();
+  bool success = loadSchema( reply.content() );
+  if ( success )
+  {
+    success = loadHierarchy();
+  }
+
+  mIsValid = success;
+}
+
+bool QgsRemoteEptPointCloudIndex::loadSchema( const QByteArray &data )
+{
+  QJsonParseError err;
```

Comment:
looks like there is quite a lot of duplicate code when there is schema loded async and in sync fashion. wondering if it is possible to reuse the code somehow?